cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -288,6 +295,50 @@ public void run() {
     }
   }
 
+  /**
+   * A runnable class to auto-adjust the allowed inter-broker partition reassignment concurrency for ongoing executions
+   * using selected broker metrics and based on additive-increase/multiplicative-decrease (AIMD) feedback control algorithm.
+   * Skips concurrency adjustment for demote operations.
+   */
+  private class ConcurrencyAdjuster implements Runnable {
+    private final int _maxPartitionMovementsPerBroker;
+    private LoadMonitor _loadMonitor;
+
+    public ConcurrencyAdjuster() {
+      _maxPartitionMovementsPerBroker = _config.getInt(ExecutorConfig.CONCURRENCY_ADJUSTER_MAX_PARTITION_MOVEMENTS_PER_BROKER_CONFIG);
+      _loadMonitor = null;
+    }
+
+    public void setLoadMonitor(LoadMonitor loadMonitor) {
+      _loadMonitor = loadMonitor;
+    }
+
+    private boolean canRefreshConcurrency() {
+      return _concurrencyAdjusterEnabled && _executorState.state() == ExecutorState.State.INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS
+             && !_isLatestExecutionDemote && _loadMonitor != null;
+    }
+
+    private void refreshConcurrency() {
+      if (canRefreshConcurrency()) {
+        Integer recommendedConcurrency = ExecutionUtils.recommendedConcurrency(_loadMonitor.currentBrokerMetricValues(),
+                                                                               _executionTaskManager.interBrokerPartitionMovementConcurrency(),
+                                                                               _maxPartitionMovementsPerBroker);
+        if (recommendedConcurrency != null) {
+          setRequestedInterBrokerPartitionMovementConcurrency(recommendedConcurrency);
+        }
+      }
+    }
```

Comment:
I don't see a lock or other synchronization mechanisms in these functions. But since they are accessing the some shared members, will there by any race condition?

Reply:
There is indeed a case that may potentially override the initial request concurrency by `ConcurrencyAdjuster` thread. Updated the code to prevent this -- thanks!

Note that the check on `_executorState.state()` is just an optimization to minimize updates to inter-broker partition movement concurrency. It has no affect to set it in any other execution state. Hence, no strict synchronization is needed.